### PR TITLE
Add float arg to build_assemblies.py

### PR DESCRIPTION
**Pull request by AlmightyLaxz**
`./modules/mono/build_scripts/build_assemblies.py` currently has no way to add the REAL_T_IS_DOUBLE DefineConstant (#65246)

This PR adds a `--float` argument to `build_assemblies.py` allowing easier compiling of the NuGet packages with double precision support.

If the float argument is 64, the NuGet packages are built with REAL_T_IS_DOUBLE added to the DefineConstants.
```sh
./modules/mono/build_scripts/build_assemblies.py --godot-output-dir ./bin \
    --push-nupkgs-local ~/MyLocalNugetSource --float=64
```

### CODE REVIEW

File: modules/mono/build_scripts/build_assemblies.py [hunk 1]
```diff
@@ -300,6 +306,7 @@ def main():
     parser.add_argument("--godot-platform", type=str, default="")
     parser.add_argument("--mono-prefix", type=str, default="")
     parser.add_argument("--push-nupkgs-local", type=str, default="")
+    parser.add_argument("--float", type=int, default=32, help="Set to 64 if Godot is compiled with double precision (REAL_T_IS_DOUBLE)")
```
[the remark this answers is not in the document]
Comment (neikeq) [reply modules/mono/build_scripts/build_assemblies.py:309]: The choices change makes sense, as I think argparse can reject invalid inputs that way.
What would `default` be here?

File: modules/mono/glue/GodotSharp/GodotSharp/GodotSharp.csproj [hunk 2]
```diff
@@ -35,7 +35,7 @@
     </None>
   </ItemGroup>
   <PropertyGroup>
-    <DefineConstants>$(DefineConstants);GODOT</DefineConstants>
+    <DefineConstants>$(DefineConstants);$(ExtraDefineConstants);GODOT</DefineConstants>
```
[the remark this answers is not in the document]
Comment (neikeq) [reply modules/mono/glue/GodotSharp/GodotSharp/GodotSharp.csproj:38]: Yes, a `GodotRealTIsDouble` property would be better. Even if the `Godot` prefix is not needed, let's keep it for consistency.

[the remark this answers is not in the document]
Comment (neikeq) [reply modules/mono/glue/GodotSharp/GodotSharp/GodotSharp.csproj:38]: Perhaps we may want to change the property name to `GodotFloat64` or something else?

File: modules/mono/build_scripts/build_assemblies.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (raulsntos) [reply modules/mono/build_scripts/build_assemblies.py:309]: The editor only checks if `float` is 64 to define `REAL_T_IS_DOUBLE` so both `default` and `32` are essentially the same but I guess the idea is in the future we could change the default to be the same as `64`.

File: modules/mono/build_scripts/build_assemblies.py [hunk 3]
```diff
@@ -277,6 +281,8 @@ def build_all(msbuild_tool, module_dir, output_dir, godot_platform, dev_debug, p
     args = ["/restore", "/t:Build", "/p:Configuration=Release"]
     if push_nupkgs_local:
         args += ["/p:ClearNuGetLocalCache=true", "/p:PushNuGetToLocalSource=" + push_nupkgs_local]
+    if float_size == 64:
```
Comment (yedpodtrzitko) [inline modules/mono/build_scripts/build_assemblies.py:284]:
```suggestion
    if float_size == "64":
```
...since the parameter `--float` is now typed as`str`

File: modules/mono/build_scripts/build_assemblies.py [hunk 4]
```diff
@@ -216,6 +216,8 @@ def build_godot_api(msbuild_tool, module_dir, output_dir, push_nupkgs_local):
         args = ["/restore", "/t:Build", "/p:Configuration=" + build_config, "/p:NoWarn=1591"]
         if push_nupkgs_local:
             args += ["/p:ClearNuGetLocalCache=true", "/p:PushNuGetToLocalSource=" + push_nupkgs_local]
+        if float_size == 64:
+            args += ["/p:ExtraDefineConstants=REAL_T_IS_DOUBLE"]
```
Comment (raulsntos) [inline modules/mono/build_scripts/build_assemblies.py:220]: You forgot to update it here.

```suggestion
        if float_size == "64":
            args += ["/p:GodotFloat64=true"]
```

Comment (AlmightyLaxz) [reply modules/mono/build_scripts/build_assemblies.py]: Whoops, thanks

File: modules/mono/build_scripts/build_assemblies.py [hunk 5]
```diff
@@ -216,6 +216,8 @@ def build_godot_api(msbuild_tool, module_dir, output_dir, push_nupkgs_local):
         args = ["/restore", "/t:Build", "/p:Configuration=" + build_config, "/p:NoWarn=1591"]
         if push_nupkgs_local:
             args += ["/p:ClearNuGetLocalCache=true", "/p:PushNuGetToLocalSource=" + push_nupkgs_local]
+        if float_size == 64:
```
Comment (yedpodtrzitko) [inline modules/mono/build_scripts/build_assemblies.py:219]:
```suggestion
        if float_size == "64":
```

**Comment by raulsntos** [comment]
FYI, the email used to author your commit doesn't seem to be linked to your GitHub account (see the email in https://patch-diff.githubusercontent.com/raw/godotengine/godot/pull/65259.patch).

It's not a problem per se, but if you want to claim the commit as a contribution for your @AlmightyLaxz account, you can add this email as secondary email in your GitHub settings, or edit the commit to change its Git author details.

**Comment by AlmightyLaxz** [comment]
Thanks for that, I spent longer than I'd like to admit working out how to squash the review changes down to one commit and must have used something with a different git config.

File: modules/mono/build_scripts/build_assemblies.py [hunk 6]
```diff
@@ -315,12 +324,7 @@ def main():
         sys.exit(1)
 
     exit_code = build_all(
-        msbuild_tool,
-        module_dir,
-        output_dir,
-        args.godot_platform,
-        args.dev_debug,
-        args.push_nupkgs_local,
+        msbuild_tool, module_dir, output_dir, args.godot_platform, args.dev_debug, args.push_nupkgs_local, args.float
```
Comment (yedpodtrzitko) [inline modules/mono/build_scripts/build_assemblies.py:327]: `black` formatting depends on the trailing comma after last item. If the comma isnt there, it will try to put all the items on the same line. If there is a trailing comma, each item will be on a separate line. So having a trailing comma here would keep the original formatting and the diff would be smaller. Just fyi.

Comment (AlmightyLaxz) [reply modules/mono/build_scripts/build_assemblies.py]: Gotcha, thanks.

File: modules/mono/build_scripts/build_assemblies.py [hunk 7]
```diff
@@ -300,6 +306,9 @@ def main():
     parser.add_argument("--godot-platform", type=str, default="")
     parser.add_argument("--mono-prefix", type=str, default="")
     parser.add_argument("--push-nupkgs-local", type=str, default="")
+    parser.add_argument(
+        "--float", type=str, default="default", choices=["default", "32", "64"], help="Floating-point precision"
```
Comment (neikeq) [inline modules/mono/build_scripts/build_assemblies.py:310]: The current code doesn't seem to be handling the "default" value. I think it should be something like this:

```py
if args.float == "default":
    args.float = "32"
```

But let's have @raulsntos confirm this.

**Review by neikeq — COMMENTED** [review]
Can you change the property to `GodotFloat64` as well in the Sdk?:

https://github.com/godotengine/godot/blob/3fe3c4399b56e24ebef5222b25044c2a18bbf55a/modules/mono/editor/Godot.NET.Sdk/Godot.NET.Sdk/Sdk/Sdk.targets#L12-L15

File: modules/mono/build_scripts/build_assemblies.py (hunk 7 above)
Comment (yedpodtrzitko) [reply modules/mono/build_scripts/build_assemblies.py:310]: ...plus `32` isnt doing anything in the code neither, so the only meaningful value is `64`, in which case it might be just a boolean switch like `--float64`?

Comment (AlmightyLaxz) [reply modules/mono/build_scripts/build_assemblies.py:310]: I put in the default value handling as it is more explicit. Without the handling it would just not use REAL_T_IS_DOUBLE but putting it in explicitly makes it clearer if the default is going to be changed engine-wide in future.

The float value has choices as that's consistent with how it is set when building the engine with scons.
It would be less friendly to vary it from the scons parameter (`scons float=64 p=linuxbsd etc`)

Comment (akien-mga) [reply modules/mono/build_scripts/build_assemblies.py:310]: The `default` doesn't seem to be used for anything in SConstruct either, so maybe it can be changed in both place to default to `32`.

File: modules/mono/build_scripts/build_assemblies.py [hunk 8]
```diff
@@ -300,9 +306,15 @@ def main():
...
+    )
 
     args = parser.parse_args()
 
+    if args.float == "default":
+        args.float == "32"
```
Comment (neikeq) [inline modules/mono/build_scripts/build_assemblies.py:316]:
```suggestion
        args.float = "32"
```

**Review by neikeq — CHANGES_REQUESTED** [review]

File: modules/mono/build_scripts/build_assemblies.py (hunk 7 above)
Comment (raulsntos) [reply modules/mono/build_scripts/build_assemblies.py:310]: The `32` value doesn't need to set anything which results in `REAL_T_IS_DOUBLE` being unset. Since our default is the same as 32, `default` also doesn't need to do anything but I agree that making it more explicit by setting the argument to `32` is better, since now the rest of code only needs to check for the `32` or `64` values and can ignore `default` (although, just to be clear, since the way our code is written defaults to 32-bit types we don't need to do anything special in that case, only when the value is `64` do we need to set `REAL_T_IS_DOUBLE` to ensure the build uses 64-bit types instead).

I also agree that it's better to keep the argument as a string rather than boolean because that would allow us to potentially support other values in the future (e.g.: `128`) which is probably why it was implemented like this in SConstruct.

**TL;DR:** I agree with @neikeq's  and @akien-mga's suggestions.

File: modules/mono/editor/Godot.NET.Sdk/Godot.NET.Sdk/Sdk/Sdk.targets [hunk 9]
```diff
@@ -10,9 +10,9 @@
     <!--
     Define constant to determine whether the real_t type in Godot is double precision or not.
     By default this is false, like the official Godot builds. If someone is using a custom
-    Godot build where real_t is double, they can override the GodotRealTIsDouble property.
+    Godot build where real_t is double, they can override the GodotFloat64 property.
     -->
-    <DefineConstants Condition=" '$(GodotRealTIsDouble)' == 'true' ">GODOT_REAL_T_IS_DOUBLE;$(DefineConstants)</DefineConstants>
+    <DefineConstants Condition=" '$(GodotFloat64)' == '64' ">REAL_T_IS_DOUBLE;$(DefineConstants)</DefineConstants>
```
Comment (raulsntos) [inline modules/mono/editor/Godot.NET.Sdk/Godot.NET.Sdk/Sdk/Sdk.targets:15]:
```suggestion
    <DefineConstants Condition=" '$(GodotFloat64)' == 'true' ">REAL_T_IS_DOUBLE;$(DefineConstants)</DefineConstants>
```

File: modules/mono/build_scripts/build_assemblies.py (hunk 7 above)
Comment (AlmightyLaxz) [reply modules/mono/build_scripts/build_assemblies.py:310]: Actually, wouldn't it be better to remove the `default` option from both SConstruct & build_assemblies.py and just set the real default with the definition of the command-line parameters?

It looks a little weird to me to put 'default' as the default value and it just means the 'real' default value needs to be set elsewhere.

We could just do it like this in SConstruct:
```
opts.Add(EnumVariable("float", "Floating-point precision", "32", ("32", "64")))
```

**Review by neikeq — APPROVED** [review]

**Review by akien-mga — APPROVED** [review]

File: modules/mono/build_scripts/build_assemblies.py [hunk 10]
```diff
@@ -300,6 +306,9 @@ def main():
     parser.add_argument("--godot-platform", type=str, default="")
     parser.add_argument("--mono-prefix", type=str, default="")
     parser.add_argument("--push-nupkgs-local", type=str, default="")
+    parser.add_argument(
+        "--float", type=str, default="32", choices=["default", "32", "64"], help="Floating-point precision"
```
Comment (raulsntos) [inline modules/mono/build_scripts/build_assemblies.py:310]: Since we removed `default`.

```suggestion
        "--float", type=str, default="32", choices=["32", "64"], help="Floating-point precision"
```

File: modules/mono/editor/Godot.NET.Sdk/Godot.NET.Sdk/Sdk/Sdk.targets [hunk 11]
```diff
@@ -10,9 +10,9 @@
     <!--
     Define constant to determine whether the real_t type in Godot is double precision or not.
     By default this is false, like the official Godot builds. If someone is using a custom
-    Godot build where real_t is double, they can override the GodotRealTIsDouble property.
+    Godot build where real_t is double, they can override the GodotFloat64 property.
     -->
-    <DefineConstants Condition=" '$(GodotRealTIsDouble)' == 'true' ">GODOT_REAL_T_IS_DOUBLE;$(DefineConstants)</DefineConstants>
+    <DefineConstants Condition=" '$(GodotFloat64)' == 'true' ">REAL_T_IS_DOUBLE;$(DefineConstants)</DefineConstants>
```
Comment (raulsntos) [inline modules/mono/editor/Godot.NET.Sdk/Godot.NET.Sdk/Sdk/Sdk.targets:15]: @neikeq Do we want to change the constant name in the Sdk?

Comment (AlmightyLaxz) [reply modules/mono/editor/Godot.NET.Sdk/Godot.NET.Sdk/Sdk/Sdk.targets:15]: I couldn't find the old `GODOT_REAL_T_IS_DOUBLE` name in use anywhere

Comment (raulsntos) [reply modules/mono/editor/Godot.NET.Sdk/Godot.NET.Sdk/Sdk/Sdk.targets:15]: This is the Sdk, it defines the constant to be used in game projects.

Comment (AlmightyLaxz) [reply modules/mono/editor/Godot.NET.Sdk/Godot.NET.Sdk/Sdk/Sdk.targets:15]: In that case it might be good idea to maintain the GODOT prefix

Comment (neikeq) [reply modules/mono/editor/Godot.NET.Sdk/Godot.NET.Sdk/Sdk/Sdk.targets]: I don't have a strong opinion. The benefit of removing the prefix is consistency. Originally, I used the prefix to make it clear this was a Godot constant and to avoid conflicts, but it's not like there's any code out there using a constant named `REAL_T_IS_DOUBLE` other than Godot.

Comment (neikeq) [reply modules/mono/editor/Godot.NET.Sdk/Godot.NET.Sdk/Sdk/Sdk.targets]: I'll let you guys decide. We can merge once this suggestion is resolved.

**Review by raulsntos — APPROVED** [review]

**Comment by neikeq** [comment]
Thanks!
